src/axom/spin/internal/linear_bvh/vec.hpp

```diff
@@ -0,0 +1,427 @@
+// Copyright (c) 2017-2019, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+
+#ifndef AXOM_SPIN_BVH_VEC_H_
+#define AXOM_SPIN_BVH_VEC_H_
+
+#include "axom/core/Macros.hpp"
+#include "axom/core/Types.hpp"    // for fixed bitwidth types
+
+#include <assert.h>
+#include <iostream>
+
+#ifdef __CUDACC__
+#include "math.h"
+#else
+#include <math.h>
+#endif
+
+namespace axom
+{
+namespace spin
+{
+namespace internal
+{
+namespace linear_bvh
+{
+
+template<typename T, axom::int32 S>
+class Vec
+{
+
+public:
+  T m_data[S];
+
+  //
+  //  No contructors so this is a POD type
+  //
+
+  inline AXOM_HOST_DEVICE bool operator==(const Vec<T,S> &other) const
+  {
+    bool e = true;
+    for(int i = 0; i < S; ++i)
+    {
+      if(m_data[i] != other[i]) e = false;
+    }
+    return e;
+  }
+
+  template<typename TT, axom::int32 SS>
+  friend std::ostream& operator<<(std::ostream &os, const Vec<TT,SS> &vec);
+
+  inline AXOM_HOST_DEVICE void operator=(const Vec<T,S> &other)
+  {
+    for(int i = 0; i < S; ++i)
+    {
+      m_data[i] = other.m_data[i];
+    }
+  }
+
+  inline AXOM_HOST_DEVICE const T& operator[](const axom::int32 &i) const
+  {
+    assert(i > -1 && i < S);
+    return m_data[i];
+  }
+
+  inline AXOM_HOST_DEVICE T& operator[](const axom::int32 &i)
+  {
+    assert(i > -1 && i < S);
+    return m_data[i];
+  }
+
+  // scalar mult /  div
+  inline AXOM_HOST_DEVICE Vec<T,S> operator*(const T &s) const
```

Comment:
It can be a bit problematic to include the arithmetic operators since they all create temporaries. It's certainly convenient but it might be worth it to just define the `+=, ...` operators and let the user figure out exactly what temporaries need to be created. If you do decided to keep these operators maybe implement them as such
``` c++
Vec<T, S> res = *this;
res *= s;
return res;
```

Reply:
I agree. This was already there. We are not using it anywhere in the BVH code.